Add cluster-dns script

bernardjkim opened the pull request:
## Description
This PR adds a `cluster-dns.sh` script that executes before the kubelet service is started. The script writes the cluster DNS servers into the kubelet config file. 

This is a necessary change before migrating to Kubernetes 1.19. Prior to 1.19, the cluster DNS servers are configured using the  `--cluster-dns` flag. This flag is now deprecated and the servers need to be configured through the kubelet config file.

bernardjkim committed:
Configure cluster dns in kubelet config file

- Add cluster-dns.sh script that writes clusterDNS to kubelet.yaml.
- Update kube-kubelet.service to exec cluster-dns.sh before starting.

UnderGreen commented on build.assets/makefiles/master/k8s-master/cluster-dns.sh:
```diff
@@ -0,0 +1,43 @@
+fi
+
+# Remove existing clusterDNS values.
+perl -i -p0e 's/clusterDNS:(\n  - .*)*\n//g' $KUBELET_CONFIG
```
I know it is up to you which tool to use here, but `sed`, in my opinion, is better known by anyone who will support the code and can provide the same result.

bernardjkim replied:
Thats fair, I've updated the script in 6e2b26f to use `sed` instead of `perl`. I'm not too familiar with `sed`, so let me know if the command can be simplified.

UnderGreen commented on build.assets/makefiles/master/k8s-master/cluster-dns.sh:
```diff
@@ -0,0 +1,43 @@
+CLUSTER_DNS="clusterDNS:"
+while IFS=',' read -ra ADDR; do
+    for i in "${ADDR[@]}"; do
+        CLUSTER_DNS+="\n  - $i"
```
The same here - could be achieved by using sed or tee right here and don't call `echo -e` later.

bernardjkim replied:
What is the benefit of using `tee` over `echo`?

UnderGreen replied:
I mean you can use sed/tee right here on line 40 without calling echo later

bernardjkim replied:
I spent some time trying to figure out how to delete and append in the same command, but wasn't able to get it figured out. I don't want to spend too much time trying to learn how to use `sed` so I'll just leave it like this for now as long as it is producing the correct results.

knisbet replied:
I think that's fine if it's working.

bernardjkim committed:
Use tee and sed instead of perl

bernardjkim committed:
Use slightly stricter pattern

- sed command will now delete until the line starts with a non space
  character. Not sure if IPv6 can be used for DNS server addresses, but
  just to be safe.
- Remove extra newline after listing DNS servers.

knisbet commented on build.assets/makefiles/master/k8s-master/cluster-dns.sh:
```diff
@@ -0,0 +1,49 @@
+# Path to kubelet config file.
+KUBELET_CONFIG=/etc/kubernetes/kubelet.yaml
+
+# If /run/dns.env file does not exist exit without any changes.
```
I believe it's possible to race during startup, so I think you want to exit with an error so that kubelet doesn't start until the file exists. 

Although I guess to be fair the file is still listed as an env file in the unit, so maybe this shouldn't really be possible unless something really funny occurs.

bernardjkim committed:
cluster-dns.sh will exit with error if ...

- dns.env is not found or
- dns addresses is empty

bernardjkim committed:
Update regex

List of addresses may be un-indented.

bernardjkim committed:
Send stdout of tee to /dev/null
